pyccl/baryons.py

```diff
@@ -0,0 +1,137 @@
+from . import ccllib as lib
+from .base import CCLAutoreprObject
+from .pyutils import check
+from .pk2d import Pk2D
+from .base import unlock_instance
+import numpy as np
+
+
+class BaryonicEffects(CCLAutoreprObject):
+    """`BaryonicEffect` obects are used to include the imprint of baryons
+    on the non-linear matter power spectrum. Their main ingredient is a
+    method `apply_baryons` that takes in a `ccl.Pk2D` and returns another
+    `ccl.Pk2D` object that now accounts for baryonic effects.
+    """
+
+    def apply_baryons(self, cosmo, pk2d, in_place=False):
```

Comment:
I disagree. I'd rather keep both things decoupled (just like we are doing with Pk2Ds and halo model calculators).

The point of having a superclass right now, is that we will in the future have other subclasses of it. It also holds the `_new_pk` method. If you can propose a useful `__init__()`, I'm all for it, but I'd rather not use cosmo as an excuse to create one.

Reply:
First of all, the `_new_pk` method doesn't use `self` at all, so it could well have been a standalone function (if we decide to keep it).

The use-case here is not akin to `Pk2D` or the `HMCalculator`: here, we deal with parameters.

I don't see the point of keeping them decoupled. `cosmo` just holds parameters, and can be treated as such, giving `BaryonModel` the ability to update it if needed. What I propose is for `BaryonModel` to merely be a 'prescription', that needs the baryonic parameters + the cosmological parameters to work. And it acts on `pk`.

Otherwise, what is the purpose of having it in a class? We could simply move the baryonic parameters to `include_baryonic_feeback` as well, and just call it using `include_baryonic_feeback(pk, cosmo, **bar_params)`. This beats the whole purpose of a class though.

Reply:
For cleanliness, modularity, and flexibility. Two examples:
a) Likelihoods that take advantage of slow-fast split will perhaps want to set up cosmological and baryonic parameters in different parts of the code. This allows for a complete decoupling.
b) In the future, there may be other `Baryons` objects with expensive `__init__`s that would be best to run only once, and not every time the cosmology changes.
c) This formalism still allows for developers to create `Baryons` classes that must be initialised with a cosmology if necessary. I do not think it's a good idea, and definitely not for this model. The only reason why we actually use the cosmology is because the model was formulated in h-inverse units, but it could just as well have been defined without that, and the cosmology would be irrelevant.

Reply:
I think I've explained my reasons above, and at this point I would just be repeating myself.

Reply:
We took this discussion on Slack and agreed to leave `cosmo` as is.